substrate-archive/src/wasm_tracing.rs

```diff
@@ -332,3 +348,58 @@ fn parse_target(s: &str) -> (String, Level) {
 		None => (s.to_string(), Level::TRACE),
 	}
 }
+
+#[cfg(test)]
+mod tests {
+	use super::*;
+	use anyhow::Error;
+	use sc_executor::{WasmExecutionMethod, WasmExecutor};
+	use sc_executor_common::runtime_blob::RuntimeBlob;
+	use sp_io::TestExternalities;
+	use sp_wasm_interface::HostFunctions;
+	use test_common::wasm_binary_unwrap;
+
+	const TARGETS: &str = "wasm_tracing,test_wasm";
+
+	#[test]
+	fn should_collect_spans_and_events_in_wasm() -> Result<(), Error> {
+		let mut ext = TestExternalities::default();
+		let mut ext = ext.ext();
+
+		let executor = WasmExecutor::new(
+			WasmExecutionMethod::Compiled,
+			Some(8),
+			sp_io::SubstrateHostFunctions::host_functions(),
+			8,
+			None,
+		);
+
+		let span_events = Arc::new(Mutex::new(SpansAndEvents { spans: Vec::new(), events: Vec::new() }));
+		let handler = TraceHandler::new(&TARGETS, 1337, vec![0x00, 0x01], span_events.clone());
+		let (spans, events, _) = handler.scoped_trace(|| {
+			executor
+				.uncached_call(
+					RuntimeBlob::uncompress_if_needed(&wasm_binary_unwrap()[..]).unwrap(),
+					&mut ext,
+					true,
+					"test_trace_handler",
+					&[],
+				)
+				.unwrap();
+			Ok(())
+		})?;
+		assert!({
+			spans[0].name == "im_a_span"
```

Comment:
dq: where are we getting the dummy data from? ext?

Reply:
the `lib.rs` of this test-wasm crate: https://github.com/paritytech/substrate-archive/tree/insipx/refactor-tracing/test-common/test-wasm just compiles to a wasm blob. That blob is available under a constant and is executed with the default substrate `WasmExecutor` further above the code, and surrounded with the Trace subscriber

Reply:
I think I see, its from `test_trace_handler`?

Reply:
yup!